### PR TITLE
The End of the Getter era

### PR DESCRIPTION
**Pull request by Efnilite**
### Description
Removes Getter from BukkitEventValues and all other classes.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprFoodLevel.java [hunk 1]
```diff
@@ -61,14 +59,11 @@ public boolean init(final Expression<?>[] vars, final int matchedPattern, final
...
-		return get(source, new Getter<Number, Player>() {
-			@Override
-			public Number get(final Player p) {
-				if (getTime() >= 0 && e instanceof FoodLevelChangeEvent && p.equals(((FoodLevelChangeEvent) e).getEntity()) && !Delay.isDelayed(e)) {
-					return 0.5f * ((FoodLevelChangeEvent) e).getFoodLevel();
-				}
-				return 0.5f * p.getFoodLevel();
+		return get(source, p -> {
+			if (getTime() >= 0 && e instanceof FoodLevelChangeEvent && p.equals(((FoodLevelChangeEvent) e).getEntity()) && !Delay.isDelayed(e)) {
+				return 0.5f * ((FoodLevelChangeEvent) e).getFoodLevel();
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprFoodLevel.java:64]: variable names

File: src/main/java/ch/njol/skript/expressions/ExprTool.java [hunk 2]
```diff
@@ -77,10 +75,9 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 	@Override
 	protected Slot[] get(final Event e, final LivingEntity[] source) {
 		final boolean delayed = Delay.isDelayed(e);
-		return get(source, new Getter<Slot, LivingEntity>() {
+		return get(source, new Converter<>() {
 			@Override
-			@Nullable
-			public Slot get(final LivingEntity ent) {
+			public Slot convert(final LivingEntity ent) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprTool.java:80]: why not lambda

File: src/main/java/ch/njol/skript/expressions/ExprWeather.java [hunk 3]
```diff
@@ -65,14 +59,11 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
-				if (getTime() >= 0 && e instanceof WeatherEvent && w.equals(((WeatherEvent) e).getWorld()) && !Delay.isDelayed(e))
-					return WeatherType.fromEvent((WeatherEvent) e);
-				else
-					return WeatherType.fromWorld(w);
-			}
+		return get(source, w -> {
+			if (getTime() >= 0 && e instanceof WeatherEvent && w.equals(((WeatherEvent) e).getWorld()) && !Delay.isDelayed(e))
+				return WeatherType.fromEvent((WeatherEvent) e);
+			else
+				return WeatherType.fromWorld(w);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprWeather.java:66]: variable names and pattern matching

**Review by sovdeeth — CHANGES_REQUESTED** [review]
only looked at a few but it applies to everything

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 4]
```diff
@@ -110,434 +114,142 @@ public BukkitEventValues() {
...
-		}, 0);
-		EventValues.registerEventValue(EntityEvent.class, Location.class, new Getter<Location, EntityEvent>() {
-			@Override
-			@Nullable
-			public Location get(final EntityEvent e) {
-				return e.getEntity().getLocation();
-			}
-		}, 0);
+		EventValues.registerEventValue(EntityEvent.class, Entity.class, EntityEvent::getEntity, 0, "Use 'attacker' and/or 'victim' in damage/death events", EntityDamageEvent.class, EntityDeathEvent.class);
+		EventValues.registerEventValue(EntityEvent.class, CommandSender.class, EntityEvent::getEntity, 0, "Use 'attacker' and/or 'victim' in damage/death events", EntityDamageEvent.class, EntityDeathEvent.class);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:177]:
```suggestion
		EventValues.registerEventValue(EntityEvent.class, Entity.class, EntityEvent::getEntity, TIME_NOW, "Use 'attacker' and/or 'victim' in damage/death events", EntityDamageEvent.class, EntityDeathEvent.class);
		EventValues.registerEventValue(EntityEvent.class, CommandSender.class, EntityEvent::getEntity, TIME_NOW, "Use 'attacker' and/or 'victim' in damage/death events", EntityDamageEvent.class, EntityDeathEvent.class);
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 5]
```diff
@@ -110,434 +114,142 @@ public BukkitEventValues() {
...
-			public Projectile get(final EntityDamageByEntityEvent e) {
-				if (e.getDamager() instanceof Projectile)
-					return (Projectile) e.getDamager();
-				return null;
-			}
-		}, 0);
+		EventValues.registerEventValue(EntityDamageEvent.class, DamageCause.class, EntityDamageEvent::getCause);
+		EventValues.registerEventValue(EntityDamageByEntityEvent.class, Projectile.class, e -> {
+			if (e.getDamager() instanceof Projectile)
+				return (Projectile) e.getDamager();
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:184]:
```suggestion
			if (e.getDamager() instanceof Projectile projectile)
				return projectile;
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Looks pretty good. In BukkitEventValues, there are cases where pattern matching, and TIME constants can be used. I highlighted one of each I noticed.

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 6]
```diff
@@ -110,434 +114,142 @@ public BukkitEventValues() {
...
-			public DamageCause get(final EntityDeathEvent e) {
-				final EntityDamageEvent ldc = e.getEntity().getLastDamageCause();
-				return ldc == null ? null : ldc.getCause();
-			}
-		}, 0);
+		EventValues.registerEventValue(EntityDeathEvent.class, ItemStack[].class, event -> event.getDrops().toArray(new ItemStack[0]));
+		EventValues.registerEventValue(EntityDeathEvent.class, Projectile.class, e -> {
+			final EntityDamageEvent ldc = e.getEntity().getLastDamageCause();
+			if (ldc instanceof EntityDamageByEntityEvent && ((EntityDamageByEntityEvent) ldc).getDamager() instanceof Projectile)
+				return (Projectile) ((EntityDamageByEntityEvent) ldc).getDamager();
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:192]:
```suggestion
			if (ldc instanceof EntityDamageByEntityEvent damageEvent && damageEvent.getDamager() instanceof Projectile projectile)
				return (Projectile) projectile;
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 7]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
-			@Override
-			public Entity get(HangingBreakEvent e) {
-				if (e instanceof HangingBreakByEntityEvent)
-					return ((HangingBreakByEntityEvent) e).getRemover();
-				return null;
-			}
-		}, 0);
+		EventValues.registerEventValue(HangingBreakEvent.class, Entity.class, e -> {
+			if (e instanceof HangingBreakByEntityEvent)
+				return ((HangingBreakByEntityEvent) e).getRemover();
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:354]:
```suggestion
			if (e instanceof HangingBreakByEntityEvent hangingEvent)
				return hangingEvent.getRemover();
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 8]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
-			@Override
-			@Nullable
-			public World get(final CommandEvent e) {
-				return e.getSender() instanceof Player ? ((Player) e.getSender()).getWorld() : null;
-			}
-		}, 0);
+		EventValues.registerEventValue(ServerCommandEvent.class, CommandSender.class, ServerCommandEvent::getSender);
+		EventValues.registerEventValue(CommandEvent.class, String[].class, CommandEvent::getArgs);
+		EventValues.registerEventValue(CommandEvent.class, CommandSender.class, CommandEvent::getSender);
+		EventValues.registerEventValue(CommandEvent.class, World.class, e -> e.getSender() instanceof Player ? ((Player) e.getSender()).getWorld() : null);
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:380]:
```suggestion
		EventValues.registerEventValue(CommandEvent.class, World.class, e -> e.getSender() instanceof Player player ? player.getWorld() : null);
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 9]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
-				Inventory invi = e.getClickedInventory(); // getInventory is WRONG and dangerous
-				if (invi == null)
-					return null;
-				int slotIndex = e.getSlot();
-
+		EventValues.registerEventValue(InventoryClickEvent.class, Player.class, e -> e.getWhoClicked() instanceof Player ? (Player) e.getWhoClicked() : null);
+		EventValues.registerEventValue(InventoryClickEvent.class, World.class, e -> e.getWhoClicked().getWorld());
+		EventValues.registerEventValue(InventoryClickEvent.class, ItemStack.class, e -> {
+			if (e instanceof CraftItemEvent)
+				return ((CraftItemEvent) e).getRecipe().getResult();
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:396]:
```suggestion
			if (e instanceof CraftItemEvent craftEvent)
				return craftEvent.getRecipe().getResult();
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 10]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
-			@SuppressWarnings("null")
-			@Override
-			@Nullable
-			public Slot get(final InventoryClickEvent e) {
-				Inventory invi = e.getClickedInventory(); // getInventory is WRONG and dangerous
-				if (invi == null)
-					return null;
-				int slotIndex = e.getSlot();
-
+		EventValues.registerEventValue(InventoryClickEvent.class, Player.class, e -> e.getWhoClicked() instanceof Player ? (Player) e.getWhoClicked() : null);
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:392]:
```suggestion
		EventValues.registerEventValue(InventoryClickEvent.class, Player.class, e -> e.getWhoClicked() instanceof Player player ? player : null);
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 11]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
+		});
+		EventValues.registerEventValue(InventoryClickEvent.class, Slot.class, e -> {
+			Inventory invi = e.getClickedInventory(); // getInventory is WRONG and dangerous
+			if (invi == null)
+				return null;
+			int slotIndex = e.getSlot();
+
+			// Not all indices point to inventory slots. Equipment, for example
+			if (invi instanceof PlayerInventory && slotIndex >= 36) {
+				return new ch.njol.skript.util.slot.EquipmentSlot(((PlayerInventory) invi).getHolder(), slotIndex);
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:407]:
```suggestion
			if (invi instanceof PlayerInventory playerInv && slotIndex >= 36) {
				return new ch.njol.skript.util.slot.EquipmentSlot(playerInv.getHolder(), slotIndex);
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 12]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
+			} else {
+				return new InventorySlot(invi, slotIndex, e.getRawSlot());
+			}
+		});
+		EventValues.registerEventValue(InventoryClickEvent.class, InventoryAction.class, InventoryClickEvent::getAction);
+		EventValues.registerEventValue(InventoryClickEvent.class, ClickType.class, InventoryClickEvent::getClick);
+		EventValues.registerEventValue(InventoryClickEvent.class, Inventory.class, InventoryClickEvent::getClickedInventory);
+		// InventoryDragEvent
+		EventValues.registerEventValue(InventoryDragEvent.class, Player.class, event -> event.getWhoClicked() instanceof Player ? (Player) event.getWhoClicked() : null);
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:416]:
```suggestion
		EventValues.registerEventValue(InventoryDragEvent.class, Player.class, event -> event.getWhoClicked() instanceof Player player ? player : null);
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 13]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
+		EventValues.registerEventValue(PrepareItemCraftEvent.class, Inventory.class, PrepareItemCraftEvent::getInventory);
+		EventValues.registerEventValue(PrepareItemCraftEvent.class, Player.class, e -> {
+			List<HumanEntity> viewers = e.getInventory().getViewers(); // Get all viewers
+			if (viewers.size() == 0) // ... if we don't have any
 				return null;
-			}
-		}, 0);
+			HumanEntity first = viewers.get(0); // Get first viewer and hope it is crafter
+			if (first instanceof Player) // Needs to be player... Usually it is
+				return (Player) first;
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:474]:
```suggestion
			if (first instanceof Player player) // Needs to be player... Usually it is
				return player;
```

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 14]
```diff
@@ -548,1351 +260,447 @@ public PotionEffectType get(AreaEffectCloudApplyEvent e) {
...
+		EventValues.registerEventValue(CraftItemEvent.class, String.class, e -> {
+			Recipe recipe = e.getRecipe();
+			if (recipe instanceof Keyed)
+				return ((Keyed) recipe).getKey().toString();
+			return null;
+		});
+		EventValues.registerEventValue(PrepareItemCraftEvent.class, String.class, e -> {
+			Recipe recipe = e.getRecipe();
+			if (recipe instanceof Keyed)
+				return ((Keyed) recipe).getKey().toString();
```
Comment (TheAbsolutionism) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:487]:
```suggestion
			if (recipe instanceof Keyed recipeKey)
				return recipeKey.getKey().toString();
```

**Review by TheAbsolutionism — CHANGES_REQUESTED** [review]
Would be cool if you updated these while you're at it

**Review by TheAbsolutionism — APPROVED** [review]

**Review by Burbulinis — APPROVED** [review]
thank you for your service, getters!

**Review by Moderocky — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

**Comment by Moderocky** [comment]
You almost reintroduced the ignite block issue 😬

**Pull request merged**